File: scripts/verify-chunk-hash.py
#!/usr/bin/env python3
"""Extract every recipe_translations / recipe_ingredients row from a generated
SQL chunk file and print a per-locale / per-ingredient sha256 digest covering
ALL text columns, so it can be diffed against digests computed by Postgres
after import. Read-only, does not modify the source file.

Usage:
  python3 verify-chunk-hash.py <path-to-chunk.sql>

Prints lines like:
  TRANSLATION  <recipe_id>  <locale>  <sha256>
  INGREDIENT   <recipe_id>  <sort_order>  <sha256>

Also prints ready-to-run SQL that computes the matching digests in Postgres.
"""
import re
import sys
import json
import hashlib


def parse_sql_string_literals(sql: str, start: int):
    i = start
    n = len(sql)
    values = []
    while i < n:
        while i < n and sql[i] in " \t\n":
            i += 1
        if sql[i] == "'":
            j = i + 1
            buf = []
            while j < n:
                if sql[j] == "'":
                    if j + 1 < n and sql[j + 1] == "'":
                        buf.append("'")
                        j += 2
                        continue
                    else:
                        j += 1
                        break
                else:
                    buf.append(sql[j])
                    j += 1
            values.append(("STR", "".join(buf)))
            i = j
            while i < n and sql[i] in " \t\n":
                i += 1
            if sql[i:i + 2] == "::":
                k = i + 2
                while k < n and (sql[k].isalnum() or sql[k] in "_[]"):
                    k += 1
                i = k
        elif sql[i:i + 4] == "null":
            values.append(("NULL", None))
            i += 4
        elif sql[i:i + 5] == "array":
            k = sql.index("[", i)
            depth2 = 1
            k += 1
            start_arr = k
            while depth2 > 0:
                if sql[k] == "[":
                    depth2 += 1
                elif sql[k] == "]":
                    depth2 -= 1
                k += 1
            values.append(("ARRAY", sql[start_arr:k - 1]))
            i = k
            while i < n and sql[i] in " \t\n":
                i += 1
            if sql[i:i + 2] == "::":
                kk = i + 2
                while kk < n and (sql[kk].isalnum() or sql[kk] in "_[]"):
                    kk += 1
                i = kk
        else:
            j = i
            while j < n and sql[j] not in ",)":
                j += 1
            values.append(("RAW", sql[i:j].strip()))
            i = j
        while i < n and sql[i] in " \t\n":
            i += 1
        if i < n and sql[i] == ",":
            i += 1
            continue
        elif i < n and sql[i] == ")":
            i += 1
            break
        else:
            raise ValueError(f"Unexpected char at {i}: {sql[i:i + 20]!r}")
    return values, i


def sha256_of(*parts):
    joined = "\x1f".join(p if p is not None else "" for p in parts)
    return hashlib.sha256(joined.encode("utf-8")).hexdigest()


def main():
    path = sys.argv[1]
    with open(path, "r", encoding="utf-8") as f:
        sql = f.read()

    print(f"# {path}")

    for m in re.finditer(r"insert into public\.recipe_translations \(([^)]*)\) values \(", sql):
        cols = [c.strip() for c in m.group(1).split(",")]
        start = m.end()
        values, _ = parse_sql_string_literals(sql, start)
        row = dict(zip(cols, [v[1] for v in values]))
        text_cols = ["title", "slug", "excerpt", "article", "seo_title", "seo_description"]
        steps_raw = row.get("steps") or "[]"
        steps_parsed = json.loads(steps_raw)
        steps_text = "\x1e".join(
            (step.get("text") or "") + "\x1d" + (step.get("tip") or "")
            for step in steps_parsed
        )
        digest = sha256_of(*([row.get(c) or "" for c in text_cols] + [steps_text]))
        print(f"TRANSLATION\t{row.get('recipe_id')}\t{row.get('locale')}\t{digest}")

    for m in re.finditer(r"insert into public\.recipe_ingredients \(([^)]*)\) values \(", sql):
        cols = [c.strip() for c in m.group(1).split(",")]
        start = m.end()
        values, _ = parse_sql_string_literals(sql, start)
        row = dict(zip(cols, [v[1] for v in values]))
        text_cols = ["unit_de", "unit_pl", "name_de", "name_pl", "group_name",
                     "store_hint_de", "substitute_de", "substitute_pl"]
        digest = sha256_of(*[str(row.get(c)) if row.get(c) is not None else "" for c in text_cols])
        print(f"INGREDIENT\t{row.get('recipe_id')}\t{row.get('sort_order')}\t{digest}")


if __name__ == "__main__":
    main()

File: scripts/verify-chunk-hash.mjs
#!/usr/bin/env node
// One-off helper: parses a generated seed SQL chunk file and prints, for each
// `insert into public.recipe_translations (...) values (...)` statement,
// an md5 hash per text column so it can be diffed against what actually
// landed in the database after an execute_sql call. Not part of the app.
import { readFileSync } from "node:fs";
import { createHash } from "node:crypto";

const file = process.argv[2];
const dumpTarget = process.argv[3]; // optional "recipeId:locale:column"
if (!file) {
  console.error("usage: node verify-chunk-hash.mjs <path-to-chunk.sql> [recipeId:locale:column]");
  process.exit(1);
}

const sql = readFileSync(file, "utf8");

// Tokenizer for a single top-level SQL statement's parenthesized VALUES list.
// Handles standard '' escaping inside '...' string literals, which is the
// only escaping style used by the generator.
function splitTopLevelValues(valuesText) {
  const items = [];
  let depth = 0;
  let inString = false;
  let current = "";
  for (let i = 0; i < valuesText.length; i++) {
    const ch = valuesText[i];
    if (inString) {
      current += ch;
      if (ch === "'") {
        if (valuesText[i + 1] === "'") {
          current += "'";
          i++;
        } else {
          inString = false;
        }
      }
      continue;
    }
    if (ch === "'") {
      inString = true;
      current += ch;
      continue;
    }
    if (ch === "(") {
      depth++;
      if (depth === 1) {
        current = "";
        continue;
      }
    }
    if (ch === ")") {
      depth--;
      if (depth === 0) {
        items.push(current);
        continue;
      }
    }
    if (depth >= 1) current += ch;
  }
  return items;
}

// Split a values-tuple string into top-level comma-separated fields,
// respecting nested parens/casts and quoted strings.
function splitFields(tuple) {
  const fields = [];
  let depth = 0;
  let inString = false;
  let current = "";
  for (let i = 0; i < tuple.length; i++) {
    const ch = tuple[i];
    if (inString) {
      current += ch;
      if (ch === "'") {
        if (tuple[i + 1] === "'") {
          current += "'";
          i++;
        } else {
          inString = false;
        }
      }
      continue;
    }
    if (ch === "'") {
      inString = true;
      current += ch;
      continue;
    }
    if (ch === "[" || ch === "(") depth++;
    if (ch === "]" || ch === ")") depth--;
    if (ch === "," && depth === 0) {
      fields.push(current);
      current = "";
      continue;
    }
    current += ch;
  }
  fields.push(current);
  return fields.map((f) => f.trim());
}

function unquote(field) {
  const f = field.trim();
  if (f === "null") return null;
  const m = f.match(/^'([\s\S]*)'/);
  if (!m) return f; // not a string literal (number etc.)
  return m[1].replace(/''/g, "'");
}

function md5(text) {
  if (text === null) return "NULL";
  return createHash("md5").update(text, "utf8").digest("hex");
}

const stmtRegex =
  /insert into public\.recipe_translations \(([^)]*)\)\s*values\s*/gi;
let match;
const columnsCache = {};
while ((match = stmtRegex.exec(sql))) {
  const columns = match[1].split(",").map((c) => c.trim());
  const startIdx = stmtRegex.lastIndex;
  // Find the matching outer parens for the single tuple starting here.
  let depth = 0;
  let inString = false;
  let end = -1;
  for (let i = startIdx; i < sql.length; i++) {
    const ch = sql[i];
    if (inString) {
      if (ch === "'") {
        if (sql[i + 1] === "'") {
          i++;
        } else {
          inString = false;
        }
      }
      continue;
    }
    if (ch === "'") {
      inString = true;
      continue;
    }
    if (ch === "(") depth++;
    if (ch === ")") {
      depth--;
      if (depth === 0) {
        end = i;
        break;
      }
    }
  }
  const tuple = sql.slice(startIdx + 1, end); // strip outer ( )
  const fields = splitFields(tuple);
  const byCol = {};
  columns.forEach((c, idx) => {
    byCol[c] = unquote(fields[idx]);
  });
  const recipeId = byCol["recipe_id"];
  const locale = byCol["locale"];
  if (dumpTarget) {
    const [wantId, wantLocale, wantCol] = dumpTarget.split(":");
    if (recipeId === wantId && locale === wantLocale) {
      if (wantCol === "steps_concat") {
        const steps = JSON.parse(byCol["steps"]);
        process.stdout.write(
          steps.map((s) => `${s.text}\u0001${s.tip}`).join("\u0002")
        );
      } else {
        process.stdout.write(byCol[wantCol] ?? "");
      }
    }
    continue;
  }
  console.log(`recipe_id=${recipeId} locale=${locale}`);
  for (const col of [
    "title",
    "slug",
    "excerpt",
    "steps",
    "article",
    "seo_title",
    "seo_description",
  ]) {
    const val = byCol[col];
    console.log(`  ${col}: len=${val === null ? 0 : val.length} md5=${md5(val)}`);
  }
}
